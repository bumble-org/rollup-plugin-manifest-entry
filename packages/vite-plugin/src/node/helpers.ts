import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { AcornNode, PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type {
  AcornIdentifier,
  AcornLiteral,
  AcornMemberExpression,
  AcornTemplateElement,
  ManifestFiles,
} from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export const isPresent = <T>(x: T | null | undefined): x is T =>
  x !== null && typeof x !== 'undefined'

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(isString(manifest.icons) ? [manifest.icons] : manifest.icons ?? {}) as string[],
    Object.values(isString(manifest.action?.default_icon) ? [manifest.action?.default_icon] : manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  let webAccessibleResources: string[] = []
  if (manifest.web_accessible_resources) {
    const resources = await Promise.all(
      manifest.web_accessible_resources
        .flatMap(({ resources }) => resources!)
        .map(async (r) => {
          // don't copy node_modules, etc
          if (['*', '**/*'].includes(r)) return undefined
          if (fg.isDynamicPattern(r)) return fg(r, options)
          return r
        }),
    )
    webAccessibleResources = resources.flat().filter(isString)
  }

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
    webAccessibleResources,
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_url_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
    manifest.sandbox?.pages,
  ]
    .flat()
    .filter(isString)
    .map((s) => s.split('#')[0])
    .sort()
  return [...new Set(files)]
}

export function isMemberExpression(n: AcornNode): n is AcornMemberExpression {
  return n.type === 'MemberExpression'
}

export function isLiteral(n: AcornNode): n is AcornLiteral {
  return n.type === 'Literal'
}

export function isIdentifier(n: AcornNode): n is AcornIdentifier {
  return n.type === 'Identifier'
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)

  let literal: AcornLiteral | undefined
  let templateElement: AcornTemplateElement | undefined

  simple(tree, {
    Literal(node) {
      literal = node as AcornLiteral
    },
    TemplateElement(node) {
      templateElement = node as AcornTemplateElement
    },
  })

  let manifestJson: string | undefined = literal?.value
  if (!manifestJson) manifestJson = templateElement?.value?.cooked
  if (!manifestJson) throw new Error('unable to parse manifest code')

  let result: ManifestV3 = JSON.parse(manifestJson)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}

/**
 * [Strip paths for `web_accessible_resources`'s `matches` Â· Issue
 * #282](https://github.com/crxjs/chrome-extension-tools/issues/282)
 */
export const stubMatchPattern = (pattern: string): string => {
  /**
   * Allow <all_urls> in matches section. 
   * [Issue #459](https://github.com/crxjs/chrome-extension-tools/issues/459)
   */
  if (pattern === "<all_urls>") {
    return pattern;
  }
  const [schema, rest] = pattern.split('://')
  const [origin, pathname] = rest.split('/')
  const root = `${schema}://${origin}`
  return pathname ? `${root}/*` : root
}
