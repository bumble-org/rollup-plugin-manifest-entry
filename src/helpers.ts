import { OutputAsset, OutputChunk, OutputBundle } from 'rollup'
import prettier from 'prettier'
import { CheerioFile } from './html-inputs'
import { ChromeExtensionManifest } from './manifest'

export const not = <T>(fn: (x: T) => boolean) => (x: T) => !fn(x)

export function isChunk(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'chunk'
}

export function isAsset(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'asset'
}

export function isString(x: any): x is string {
  return typeof x === 'string'
}

export function isJsonFilePath(x: any): x is string {
  return isString(x) && x.endsWith('json')
}

export const formatHtml = ($: CheerioFile) =>
  prettier.format($.html(), { parser: 'html' })

/**
 * Update the manifest source in the output bundle
 */
export const updateManifest = (
  updater: (
    manifest: ChromeExtensionManifest,
  ) => ChromeExtensionManifest,
  bundle: OutputBundle,
  handleError?: (message: string) => void,
): OutputBundle => {
  try {
    const manifestKey = 'manifest.json'
    const manifestAsset = bundle[manifestKey] as OutputAsset

    if (!manifestAsset) {
      throw new Error(
        'No manifest.json in the rollup output bundle.',
      )
    }

    const manifest = JSON.parse(
      manifestAsset.source as string,
    ) as ChromeExtensionManifest

    const result = updater(manifest)

    manifestAsset.source = JSON.stringify(result, undefined, 2)
  } catch (error) {
    if (handleError) {
      handleError(error.message)
    } else {
      throw error
    }
  }

  return bundle
}
