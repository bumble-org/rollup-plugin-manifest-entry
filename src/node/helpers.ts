import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type { AcornLiteral, ManifestFiles } from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(manifest.icons ?? {}) as string[],
    Object.values(manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_settings_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
  ]
    .flat()
    .filter(isString)
    .sort()
  return [...new Set(files)]
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)
  let literal: AcornLiteral | undefined
  simple(tree, {
    Literal(n) {
      literal = n as AcornLiteral
    },
  })
  if (!literal) throw new Error('unable to parse manifest code')
  let result: ManifestV3 = JSON.parse(literal.value)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}
