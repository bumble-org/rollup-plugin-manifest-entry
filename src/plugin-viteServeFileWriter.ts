import {
  OutputOptions,
  Plugin,
  RollupWatcher,
  RollupWatchOptions,
  watch,
} from 'rollup'
import { from } from 'rxjs'
import { resolveConfig, ViteDevServer } from 'vite'
import { interpret } from 'xstate'
import { format } from './helpers'
import {
  machine,
  model,
} from './plugin-viteServeFileWriter.machine'
import { combinePlugins, isRPCE } from './plugin_helpers'
import { CrxPlugin } from './types'
import {
  narrowEvent,
  useConfig,
  waitForState,
} from './xstate_helpers'

/** The service is used by multiple exports */
const service = interpret(
  machine.withConfig({
    services: {
      fileWriter:
        ({ plugins: watchPlugins, server }) =>
        (send) => {
          const {
            config: { configFile, mode, build },
          } = server!
          let watcher: RollupWatcher | undefined
          ;(async () => {
            // get build config
            console.log(
              `> Resolving CRX watch config from "${configFile}"`,
            )

            const {
              plugins: buildPlugins,
              build: { rollupOptions, watch: watchOptions },
            } = await resolveConfig(
              { configFile, build },
              'build',
              mode,
            )

            // replace build crx plugins with watch crx plugins
            const plugins = combinePlugins(
              (buildPlugins as CrxPlugin[]).filter(
                ({ crx }) => !crx,
              ),
              watchPlugins,
            )

            const watchRPCE = watchPlugins.find(isRPCE)!
            plugins.forEach((p, i) => {
              // Vite and Jest resolveConfig behavior is different
              // In Vite, the config module is imported twice as two different modules
              // In Jest, not only is the config module the same,
              //   the same plugin return value is used ¯\_(ツ)_/¯
              // So we need to check that RPCE is not the same object
              if (isRPCE(p) && watchRPCE !== p) {
                // replace build RPCE with watch RPCE
                ;(plugins as CrxPlugin[])[i] = watchRPCE
                // stop the build RPCE service
                p.api.service.stop()
              }
            })

            const options: RollupWatchOptions = {
              ...rollupOptions,
              // The context should not be touched here
              // The hybrid output plugin will change it
              context: 'this',
              output: {
                ...(rollupOptions?.output as OutputOptions),
                dir: build.outDir,
              },
              plugins: plugins as Plugin[],
              watch: {
                ...watchOptions,
                chokidar: {
                  ignored: [
                    '**/node_modules/**',
                    '**/.git/**',
                    ...(watchOptions?.chokidar?.ignored || []),
                  ],
                  ignoreInitial: true,
                  ignorePermissionErrors: true,
                  ...watchOptions?.chokidar,
                },
              },
            }
            watcher = watch(options)

            watcher.on('event', async (event) => {
              if (event.code === 'BUNDLE_END') {
                send(model.events.BUNDLE_END(event))
              } else if (event.code === 'BUNDLE_START') {
                send(model.events.BUNDLE_START(event))
              } else if (event.code === 'ERROR') {
                send(model.events.ERROR(event.error))
              }
            })
          })().catch((err) => send(model.events.ERROR(err)))

          return () => watcher?.close()
        },
      waitForServer: ({ server }) =>
        from(
          new Promise<ViteDevServer>((resolve, reject) => {
            if (!server?.httpServer)
              reject(
                new Error(
                  `vite httpServer is ${typeof server?.httpServer}`,
                ),
              )

            server?.httpServer?.once('listening', resolve)
          })
            .then(model.events.SERVER_READY)
            .catch((error) =>
              model.events.ERROR(error, 'waitForServer'),
            ),
        ),
    },
  }),
  { devTools: true },
)

/**
 * Writes extension files during Vite serve.
 *
 * The Vite dev server doesn't write files to the disk,
 * which is a requirement for Chrome Extension development.
 *
 * This plugin uses an instance of Rollup Watch to
 * fulfill this requirement. We'll use a plugin to load
 * files from the dev server, and then write them to disk.
 * This keeps configuration consistent and simple.
 *
 * Since we're still using the Vite server to load files,
 * Vite will call certain hooks (resolveId, load, transform).
 * These hooks should be excluded from Rollup Watch
 * to avoid duplicate hook calls.
 */
export function viteServeFileWriter(): CrxPlugin {
  let isViteServe: boolean

  return {
    name: 'vite-serve-file-writer',
    crx: true,
    enforce: 'pre',
    config(config, { command }) {
      isViteServe = command === 'serve'
    },
    configResolved(config) {
      if (!isViteServe) return

      useConfig(service, {
        actions: {
          handleBundleStart(context, _event) {
            narrowEvent(_event, 'BUNDLE_START')
            console.log('Building Chrome Extension files...')
          },
          handleBundleEnd(context, _event) {
            const { event } = narrowEvent(_event, 'BUNDLE_END')
            console.log(
              `Build completed in ${event.duration} ms`,
            )
          },
          handleError(context, event) {
            const { error } = narrowEvent(event, 'ERROR')
            if (error.message?.includes('is not exported by')) {
              console.warn(format`Could not complete bundle because Vite did not pre-bundle a dependency.
              You may need to add this dependency to your Vite config under \`optimizeDeps.include\`.`)
            }

            console.error(error)
          },
          handleFatalError({ server }, event) {
            const { error } = narrowEvent(event, 'ERROR')
            console.error(error)
            server?.close()
            service.stop()
          },
        },
      })

      service.start()

      const watchPlugins: CrxPlugin[] = []
      config.plugins.forEach((servePlugin: CrxPlugin, i) => {
        if (isRPCE(servePlugin)) {
          // RPCE should only run in the inner Rollup Watch hooks
          watchPlugins.push(servePlugin)
          // RPCE should not run in ViteDevServer hooks
          ;(config.plugins as CrxPlugin[])[i] = {
            name: 'chrome-extension-replacement',
          }
          return
        }

        // Don't touch other non-crx plugins
        if (!servePlugin.crx) return

        // Shallow clone crx plugins for Rollup Watch
        // b/c we're going to remove some hooks from the originals
        const watchPlugin = { ...servePlugin }
        watchPlugins.push(watchPlugin)

        // These hooks should only run in watch mode for CRX plugins
        delete servePlugin.options
        delete servePlugin.buildStart
      })
      service.send(model.events.PLUGINS(watchPlugins))
    },
    configureServer(server) {
      if (!isViteServe) return

      if (service.initialized) {
        service.send(model.events.SERVER(server))
      }
    },
  }
}

/**
 * For use in tests. Resolves when file write operation is complete.
 */
export const filesReady = () =>
  waitForState(service, (state) => {
    if (state.event.type === 'ERROR') throw state.event.error
    return state.matches({ watching: 'ready' })
  })
