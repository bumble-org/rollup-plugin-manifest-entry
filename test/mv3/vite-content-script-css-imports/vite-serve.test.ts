import { filesReady } from '$src/plugin-viteServeFileWriter'
import { jestSetTimeout } from '$test/helpers/timeout'
import fs from 'fs-extra'
import glob from 'glob'
import path from 'path'
import { createServer, ViteDevServer } from 'vite'

jestSetTimeout(30000)

const outDir = path.join(__dirname, 'dist-serve')

let devServer: ViteDevServer
beforeAll(async () => {
  await fs.remove(outDir)

  devServer = await createServer({
    configFile: path.join(__dirname, 'vite.config.ts'),
    envFile: false,
    build: { outDir },
  })
})

afterAll(async () => {
  await devServer.close()
})

test('writes entry points to disk', async () => {
  expect(fs.existsSync(outDir)).toBe(false)

  await Promise.all([devServer.listen(), filesReady()])

  expect(fs.existsSync(outDir)).toBe(true)

  const cssFiles = glob.sync(`${outDir}/assets/*.css`, {
    cwd: outDir,
  })
  expect(cssFiles.length).toBe(1)

  const [styles] = cssFiles
  const manifest = 'manifest.json'
  const content = 'content.js'

  const manifestPath = path.join(outDir, manifest)
  const manifestSource = await fs.readJson(manifestPath)

  expect(manifestSource).toMatchObject({
    content_scripts: [
      {
        matches: ['http://*/*', 'https://*/*'],
        js: ['content.js'],
        css: [styles],
      },
    ],
  })

  const contentPath = path.join(outDir, content)
  const contentSource = await fs.readFile(contentPath, 'utf8')
  expect(contentSource).toMatchSnapshot(content)

  const stylesPath = path.join(outDir, styles)
  const stylesSource = await fs.readFile(stylesPath, 'utf8')
  expect(stylesSource).toMatchSnapshot(styles)
})
