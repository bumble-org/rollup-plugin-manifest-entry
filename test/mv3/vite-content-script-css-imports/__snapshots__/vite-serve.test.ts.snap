// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes entry points to disk 1`] = `
Object {
  "background": Object {
    "service_worker": "crx-hmr-service-worker.js",
    "type": "module",
  },
  "content_scripts": Array [
    Object {
      "css": Array [
        "assets/content-89281590.css",
      ],
      "js": Array [
        "runtime-reloader--content-script.js",
        "assets/content.esm-wrapper-a8bec302.js",
      ],
      "matches": Array [
        "http://*/*",
        "https://*/*",
      ],
    },
  ],
  "description": "DEVELOPMENT build with simple auto-reloader
[6:00:00 PM] Waiting for changes...",
  "manifest_version": 3,
  "name": "Test",
  "permissions": Array [],
  "version": "1.0.0",
  "version_name": "2021-12-07T00:00:00.000Z",
  "web_accessible_resources": Array [
    Object {
      "matches": Array [
        "http://*/*",
        "https://*/*",
      ],
      "resources": Array [
        "runtime-reloader--content-script.js",
      ],
    },
    Object {
      "matches": Array [
        "http://*/*",
        "https://*/*",
      ],
      "resources": Array [
        "content.js",
      ],
    },
  ],
}
`;

exports[`writes entry points to disk: assets/content-89281590.css 1`] = `
"body {
  color: red;
}
"
`;

exports[`writes entry points to disk: content.js 1`] = `
"var styles = \\"\\";
console.log(\\"content.ts\\");
"
`;
